### DISCUSSION TITLE
An `->only()` method in query builder.

### DISCUSSION THREAD

**Discussion by BrekiTomasson** [laravel/framework#43677] [Ideas]
What I'm suggesting here is for an `->only()` method (or something with a better name, obviously) which returns the result of the query, but **only if there is a single result**. If the query, as stated at that point, has more than one result, it would throw an Exception. Think of it as a `->firstOrFail()` but with the added check of requiring there to be only a single result.

Currently (to the best of my knowledge, at least), the only way to do this is in multiple steps, first you `->get()` the results, then you `->count()` the results, and if the count is not exactly one (so if it is zero or 2+), you throw an exception manually, otherwise you continue as intended.

This would be used in scenarios where `->firstOrFail()` still runs the risk of grabbing the wrong item if the query is not clear enough or if there are too many moving parts involved to be sure that you've narrowed down the search properly. It would also more clearly show where you have potential issues in your database, such as when you have a column that _should_ be `->unique()` in your migration but isn't.

Examples of use:

```php
> Country::where('alpha2', 'US')->only()
// Returns a single entry from the Country model.

> User::where('email', 'like', '%\.se')->only()
// Throws an exception because more than one user have a `.se` email address.
```

**Comment by Sophist-UK** [comment]
I would think that `onlyOrFail()` would be the most appropriate Laravel-style name for the method you are proposing. There could also be an `only()` method that returned `null` if there were zero or >= two results. 
```
> Country::where('alpha2', 'US')->only()
> Country::where('alpha2', 'US')->onlyOrFail()
// Returns a single entry from the Country model.

> User::where('email', 'like', '%\.se')->onlyOrFail()
// Throws an exception because more than one user has a `.se` email address.

> User::where('email', 'like', '%\.se')->only()
// Returns null because more than one user has a `.se` email address.
```

**Reply by BrekiTomasson** [reply]
That works for me, yeah! You're right that it's a far more Laravel-style naming convention to put it that way.

**Comment by BrekiTomasson** [comment]
I just discovered that Laravel already has something very much like this implemented already, in the `->sole()` Builder method. It returns a model if only one result can be found, throwing exceptions if nothing is found or if two or more results are found.
